File: voghion_flutter_module/packages/vg_native_mix_flutter/ios/Classes/VgNativeMixFlutterPlugin.swift
import Flutter
import UIKit


public class VgNativeMixFlutterPlugin: NSObject, FlutterPlugin {
    
    public  var delegates:[VgFlutterMixProtocol?] = []

    public static let name = "VgNativeMixFlutterPlugin"
    public static let closeFlutterPage = "closeFlutterPage"
    public static let openNativePage = "openNativePage"
    
    
    public static var shared:VgNativeMixFlutterPlugin?
    public var methodChannel:FlutterMethodChannel?

    
  public static func register(with registrar: FlutterPluginRegistrar) {
      let channel = FlutterMethodChannel(name: "vg_native_mix_flutter", binaryMessenger: registrar.messenger())
      let instance = VgNativeMixFlutterPlugin()
      shared = instance
      shared?.methodChannel = channel
      registrar.addMethodCallDelegate(instance, channel: channel)
  }
    
    public func addDelegate(_ delgate:  VgFlutterMixProtocol){
        self.delegates.append(delgate)
    }
    public func removeLastDelegate(){
        if self.delegates.isEmpty {
            return
        }
        self.delegates.removeLast()
    }
    func parseCallArgs(arguments:Any?) ->VgParseFlutterCallArgs {
        let parseArgs = VgParseFlutterCallArgs()
        var pageName = ""
        var args:Any? = nil
        if let _arguments:Dictionary<String,Any>? = arguments as? Dictionary<String, Any>? {
            pageName = _arguments?["routeName"] as? String ?? "";
            args = _arguments?["args"]
        }
        parseArgs.routeName = pageName
        parseArgs.args = args
        return parseArgs
    }
   
  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getPlatformVersion":
      result("iOS " + UIDevice.current.systemVersion)
    case VgNativeMixFlutterPlugin.closeFlutterPage:
        for dl in self.delegates {
            if dl != nil && dl?.pageIsVisible() ?? false {
                let parseArgs = parseCallArgs(arguments: call.arguments)
                dl?.closeFlutterPage(routeName: parseArgs.routeName, args: parseArgs.args)
            }
        }
        result(true)
    case VgNativeMixFlutterPlugin.openNativePage:
        for dl in self.delegates {
            if dl != nil && dl?.pageIsVisible() ?? false {
                let parseArgs = parseCallArgs(arguments: call.arguments)
                dl?.openNativePage(routeName: parseArgs.routeName, args: parseArgs.args)
            }
        }
        result(true)
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}

File: voghion_flutter_module/packages/vg_native_mix_flutter/android/src/main/kotlin/com/voghion/vg_native_mix_flutter/VgNativeMixFlutterPlugin.kt
package com.voghion.vg_native_mix_flutter

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** VgNativeMixFlutterPlugin */
class VgNativeMixFlutterPlugin: FlutterPlugin, MethodCallHandler {
  companion object {
    const val plugiNname = "VgNativeMixFlutterPlugin"
    const val closeFlutterPage = "closeFlutterPage"
    const val openNativePage = "openNativePage"
    lateinit var shared: VgNativeMixFlutterPlugin
  }
  lateinit var methodChannel : MethodChannel
  var delegates: MutableList<VgFlutterMixInterface> = mutableListOf()

  fun addDelegate(delegate:VgFlutterMixInterface){
    this.delegates.add(delegate)
  }
  fun removeLastDelegate(){
    if (this.delegates.isEmpty()) {
      return
    }
    val  idx = this.delegates.lastIndex
    this.delegates.removeAt(idx)
  }
  fun parseCallArgs(arguments: Any?):VgParseFlutterCallArgs {
    val parseArgs = VgParseFlutterCallArgs()
    var routeName = ""
    var args: Map<String, Any?>? = null
    val  argus: HashMap<String,Any>? = arguments as? HashMap<String, Any>?
    routeName = argus?.get("routeName") as? String ?: ""
    args = argus?.get("args") as? HashMap<String, Any>?
    parseArgs.routerName = routeName
    parseArgs.args = args
    return parseArgs
  }
  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "vg_native_mix_flutter")
    methodChannel.setMethodCallHandler(this)
    shared = this
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when(call.method) {
      "getPlatformVersion" -> {
        result.success("Android ${android.os.Build.VERSION.RELEASE}")
      }
      closeFlutterPage -> {
        for (dl in this.delegates) {
          if (dl.pageIsVisible()) {
            val parseCallArgs = parseCallArgs(call.arguments)
            dl.closeFlutterPage(parseCallArgs.routerName, args = parseCallArgs.args)
          }
        }
        result.success(true)
      }
      openNativePage -> {
        for (dl in this.delegates) {
          if (dl.pageIsVisible()) {
            val parseCallArgs = parseCallArgs(call.arguments)
            dl.openNativePage(parseCallArgs.routerName, args = parseCallArgs.args)
          }
        }
        result.success(true)
      }
      else ->  {
        result.notImplemented()
      }
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    methodChannel.setMethodCallHandler(null)
  }
}
